app/src/main/java/org/wikipedia/analytics/SuggestedEditsFunnel.java

```diff
@@ -141,40 +148,75 @@ public void failure(Action action) {
             statsCollection.addCaptionStats.failures++;
         } else if (action == TRANSLATE_CAPTION) {
             statsCollection.translateCaptionStats.failures++;
+        } else if (action == ADD_IMAGE_TAGS) {
+            statsCollection.machineImageTagStats.failures++;
         }
     }
 
     public void helpOpened() {
         helpOpenedCount++;
     }
 
-    public void contributionsOpened() {
-        contributionsOpenedCount++;
+    public void log() {
+        try {
+            JSONObject jsonObject = new JSONObject(GsonUtil.getDefaultGson().toJson(statsCollection));
+            removeStatsWithNoValues(jsonObject);
+            log(
+                    "edit_tasks", jsonObject,
+                    "help_opened", helpOpenedCount,
+                    "source", (invokeSource.getName())
+            );
+        } catch (JSONException e) {
+            e.printStackTrace();
+        }
     }
 
-    public void log() {
-        log(
-                "edit_tasks", GsonUtil.getDefaultGson().toJson(statsCollection),
-                "help_opened", helpOpenedCount,
-                "scorecard_opened", contributionsOpenedCount,
-                "source", (invokeSource.getName())
-        );
+    private void removeStatsWithNoValues(JSONObject editTasksJSONObject) {
+        ArrayList<String> editTaskKeysToRemove = new ArrayList<>();
```

Comment:
I'm afraid this is way too complex for what it needs to be.
Remember that `Gson` is a very powerful and versatile library.  It would have been worthwhile to check whether it provides built-in functionality to do what we need here.  I would estimate that this is achievable with maybe 3 or 4 lines of code, plus a little boilerplate.

Reply:
The TypeAdapter way is exponentially better!!!🙌